barcode/templates/barcode/baseextra.html

```diff
@@ -0,0 +1,98 @@
+{% extends "baseextra.html" %}
+{{ super() }}
+
+{% if detect_barcode_scan %}
+  <script type="text/javascript">
+      var getBadgeNumFromBarcode = function (barcode, callback) {
+          barcode = ('' + barcode).trim();
+          callback = callback || function () {
+              };
+
+          var regexpUUID = /^[0-9a-f]{8}-?[0-9a-f]{4}-?[1-5][0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$/i;
+
+          // Remove non-word characters bracketing the barcode if they match
+          // Some scanners add these but we don't want them
+          var first = barcode.substr(0, 1),
+              last = barcode.substr(barcode.length - 1);
+          if (first === last && last.match(/\W/g)) {
+              barcode = barcode.substr(1, barcode.length - 2);
+          }
+
+          // Bypass decryption if it's a number or a UUID
+          if ($.isNumeric(barcode) || barcode.match(regexpUUID)) {
+              callback({'badge_num': barcode});
+              return;
+          }
+
+          $.ajax({
+              method: 'POST',
+              url: '../barcode/get_badge_num_from_barcode',
+              data: {
+                  barcode: barcode,
+                  csrf_token: csrf_token
+              },
+              success: function (response, status) {
+                  if (response && response['badge_num'] === -1) {
+                      response['error'] = response['message'];
+                  }
+                  callback(response);
+              },
+              error: function (response, status, statusText) {
+                  callback({'error': 'Could not decode barcode: ' + statusText});
+              }
+          });
+      };
+
+      // Override these per-page as necessary
+
+      // If there's one and only one badge_num field on the page, use that
+      // Otherwise use the currently-selected field
+      if ($('input[name=badge_num]').length === 1) {
+          var $badgeNum = $('input[name=badge_num]');
+      } else {
+          var $badgeNum = $(document.activeElement);
+      }
+
+      var scannerOnError = function () {
+          $badgeNum.focus();
+      };
+
+      var scannerOnReceive = function (event) {
+          if (event.key === '~' || event.key === '\\') {
+              event.preventDefault();
+              $badgeNum.blur();
+          }
+      };
+
+      var scannerOnComplete = function (barcode, qty) {
+          getBadgeNumFromBarcode(barcode, function (response) {
+              $badgeNum.focus();
+              if (response) {
+                  if (response['error']) {
+                      toastr.error(response['error'], '', {timeOut: 3000});
+                  } else if (response['badge_num']) {
+                      var badgeNum = response['badge_num'];
+                      if ($badgeNum.is(":text")) {
```

Comment:
This is a safety check added in case a non-text input had focus, since we don't want to change the values of anything else lest we cause unexpected behavior. This does mean that if someone has a non-text input field focused, and there's no or more than one 'badge_num' field, scanning wouldn't do anything -- but that would be true if none of this code existed, too, or if you tried to type a badge number by hand, so I figured it's not a problem.

Reply:
Our badge number fields might have type set to `number`. In fact, I kinda feel like this code is trying to do too much, or it assumes too much about the parent page that it will be included on. I think it might be more flexible to not assume anything about the containing page, and not try to modify fields at all. Not sure though.

Reply:
That's true -- I wanted a 'default' behavior, though, so that future programmers can just turn this on and overrides are only needed for special cases. Otherwise, if a page doesn't implement useBarcode, this function would just eat the scan results, right?